src/audio/buffer.c

```diff
@@ -94,6 +94,51 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
 	return buffer;
 }
 
+int buffer_resize(struct comp_buffer *buffer, uint32_t size)
+{
+	struct sof_ipc_buffer *desc = NULL;
+
+	if (!buffer) {
+		trace_buffer_error("resize error: can't to resize NULL buffer");
+		return -1;
+	}
+
+	/* validate request */
+	if (size == 0 || size > HEAP_BUFFER_SIZE) {
+		trace_buffer_error("resize error: size = %u is invalid", size);
+		return -1;
+	}
+
+	desc = &buffer->ipc_buffer;
+
+	if (buffer->alloc_size == size)
+		return 0;
+
+	if (buffer->addr)
+		rfree(buffer->addr);
+
+	buffer->addr = rballoc(RZONE_RUNTIME, desc->caps, size);
+	if (!buffer->addr) {
+		rfree(buffer);
+		trace_buffer_error("resize error: can't alloc %u bytes type %u",
+				   desc->size, desc->caps);
+		return -1;
+	}
+
+	desc->size = size;
+	buffer->alloc_size = size;
+	buffer->size = size;
+	buffer->w_ptr = buffer->addr;
+	buffer->r_ptr = buffer->addr;
+	buffer->end_addr = buffer->addr + size;
+	buffer->free = size;
+	buffer->avail = 0;
+
+	buffer_zero(buffer);
+
+	return 0;
```

Comment:
sure, I'll look into it.

Reply:
@xiulipan Thanks for the review! I'll try to answer your questions:
1) The idea is that new method set_frames() is called with pipeline params, that is walking through the whole graph. So host is getting frames from stream rate / sched (aligned with dma burst), next component will get frames from previous (if not fixed output_rate, when component will set the frames itself from its own rate) and so on...
2) yes topology changes are big, I could have left the old params there and just not use them, but 
@lgirdwood asked for removal and I myself agree it would be confusing to leave them there. And as you noticed the frame count is still in the pipeline struct that is parsed in kernel (I just set it to 0 and it is not used anywhere).